ex5/ex5.c

```diff
@@ -10,13 +10,42 @@
 
 #define MSGSIZE 16
 
-char* msg1 = "hello world #1";
-char* msg2 = "hello world #2";
-char* msg3 = "hello world #3";
+char *msg1 = "hello world #1";
+char *msg2 = "hello world #2";
+char *msg3 = "hello world #3";
 
 int main(void)
 {
     // Your code here
-    
+    int f = fork();
+    char inbuf[MSGSIZE];
+    int p[2];
+    if (pipe(p) < 0)
+    {
+        fprintf(stderr, "pipe failde\n");
+        exit(1);
+    }
+    if (f < 0)
+    {
+        printf("%s\n", "Fork faliled");
+        exit(1);
+    }
+    else if (f == 0)
+    {
+        // we are in the child branch
+        write(p[1], msg1, MSGSIZE);
+        write(p[1], msg2, MSGSIZE);
+        write(p[1], msg3, MSGSIZE);
+    }
+    else
+    {
+        // we are in the parent branch
+        for (int i = 0; i < 3; i++)
+        {
+            wait(NULL);
```

Comment:
Try taking your wait() call out of the for loop and see if that stops the program from hanging. You also might consider closing the "write" side fo the pipe with the close() call (also done outside of the for loop)although, I think that's less likely to be the issue.